extmod/modujson: Allow json.loads() to preserve dict order.

andrewleech opened the pull request:
When MICROPY_PY_UJSON_ORDERED_DICT is enabled.

Migrating away from using my previous https://github.com/micropython/micropython/pull/5323 PR to make all dict's ordered has unearthed a few other places we were relying on such functionality.

pi-anl committed:
extmod/modjson: Allow json.loads() to preserve dict order.
When MICROPY_PY_JSON_ORDERED_DICT is enabled.

Signed-off-by: Andrew Leech <andrew.leech@planetinnovation.com.au>

tve commented on extmod/modujson.c:
```diff
@@ -224,6 +224,12 @@ STATIC mp_obj_t mod_ujson_load(mp_obj_t stream_obj) {
                 break;
             case '{':
                 next = mp_obj_new_dict(0);
+                #if MICROPY_PY_UJSON_ORDERED_DICT
```
Is this a new option? Is it set anywhere?

andrewleech replied:
It is a new option, and no it's not set by default anywhere. The expectation is it'll be set in make option or board profile if someone wants it... or if it's a desieable enough feature it could be made the default

tve replied:
Hmm, doesn't that make it almost impossible to discover? Someone has to go look at the code to see "hey, look at that! I can enable MICROPY_PY_UJSON_ORDERED_DICT!". Just wondering...

andrewleech replied:
Yeah it's certainly not ideal... At the very least I should enable it on the coverage build and add a unit test - this initial PR was thrown up very quickly in the middle of a project rebase/test/release cycle on my end and I didn't want it to be forgotten.

stinos replied:
Just put in in mpconfig.h, like all other options.

andrewleech commented:
Arguably, this could be good to have enabled by `MICROPY_CPYTHON_COMPAT` rather than a new define as cpython does already load json's in order thanks to their ordered by default dict's.

As there are performance differences between regular and ordered dicts still in micropython I'm happy to have it as a separate feature enable flag for now.

I've now added the flag in mpconfig.h and added/updated unit test in coverage build.

stinos commented on tests/extmod/ujson_loads.py:
```diff
@@ -7,10 +7,20 @@
     if isinstance(o, dict):
-        print("sorted dict", sorted(o.items()))
+        if is_coverage:  # Has ordered json loads enabled
+            print("dict", o.items())
```
Just to check, this assumes the CPython used to run the test will also preserve dict order. Do all CPython versions do this, and if not: from which version on? I.e. we whould be sure this always works.

andrewleech replied:
Ah yes, that is the assumption yes. This has been the case since Python 3.6, though this aspect was written in as official in 3.7
https://docs.python.org/3.6/whatsnew/3.6.html#whatsnew36-compactdict

Arguably we should look at bringing in this same dict implementation as an option in micropython, it was shown to be significantly more efficient than the previous cpython dict and would likely be better than our current ordered dict implementation at least (certainly for very large dicts)

andrewleech replied:
... and yes this is why the new unit test is failing, cpython 3.5 is used on travis :-(

andrewleech replied:
Perhaps rather than making this a `#define` feature I should have just added the `object_pairs_hook` argument like in cpython:
https://docs.python.org/3.4/library/json.html#json.load

stinos replied:
> It would be more work to add this level of generic support rather than just added an ordered flag however

Would also be larger in code size.

Perhaps change the check so that when coverage is enabled, it just checks whether the returned dict is the same as sorted(dict).

andrewleech replied:
Checking `loaded dict == sorted dict` results the same as the comparison I've already got in place - it works on micropython with my patch and cpython 3.6+ but does not work on cpython3.5. I can put an extra try block in there to loads with `object_pairs_hook=OrderedDict` for cpython though.

stinos replied:
Ah but the idea is that CPython never runs any code for which it's behavior changes between versions, i.e. code guarded by is_coverage only gets executed by MicroPython, not CPython, and the else clause is where the expected result gets printed. Pseudocode:
```
if is_coverage:
  print(is_dict_with_same_content_and_sorted(a,b))
else:
  print(True)
```

andrewleech replied:
Ah I didn't realise that's how it would work, that's brilliant, thanks. I've updated as such.

stinos commented on extmod/modujson.c:
```diff
@@ -224,6 +224,12 @@ STATIC mp_obj_t mod_ujson_load(mp_obj_t stream_obj) {
             case '{':
                 next = mp_obj_new_dict(0);
+                #if MICROPY_PY_UJSON_ORDERED_DICT
+                // keep the locals ordered
```
Same comment as for other commit (plus this isn't about locals here)
